File: auto.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <deque>
#include <cctype>
using namespace std;

void displayCharVector(std::vector<char> arr)
{
	for (char val : arr)
	{
		std::cout << val;
	}
	std::cout << std::endl;
}
bool compare(char a, char b)
{
	if (toupper(a) == toupper(b))
	{
		return true;
	}
	return false;
}
void display(std::vector<std::pair<char, int>> val)
{
	for (auto curr : val)
	{
		std::cout << "Value = " << curr.first << " Index = " << curr.second << std::endl;
	}
}
void displayVectorDeque(std::deque<std::vector<char>> val)
{
	for (std::vector<char> elem : val)
	{
		displayCharVector(elem);
	}
}
int main()
{
	// (466551)
	std::deque<std::vector<char>> words;
	//used  <std::vector<char>> because , it's much easier to go with it

	fstream openFile("words.txt");
	std::string word = "";
	if (openFile)
	{
		while (getline(openFile, word))
		{
			std::vector<char> stringVector(word.length());
			std::copy(word.begin(), word.end(), stringVector.begin());
			words.push_back(stringVector);
			//pushing the vector to main deque
		}
	}
	else
	{
		std::cout << "Could not find the dictionary " << std::endl;
	}

	std::string input;
	std::cout << "Enter word to find ------------" << std::endl;
	std::cin >> input;
	// input = "somewhx"; // Written Just for testing

	std::vector<char> inputVector(input.length());
	std::copy(input.begin(), input.end(), inputVector.begin());
	std::deque<vector<char>>::iterator foundIterator = std::find(words.begin(), words.end(), inputVector);

	if (foundIterator == words.end())
	{
		//l
		std::cout << "No matching words found " << std::endl;
		std::cout << "Suggestions for you are \n";

		std::deque<std::vector<char>> suggestions, previousSuggestion;
		// words.size() is necessary because it won't copy if not given
		suggestions = words;
		suggestions.erase(std::remove_if(
							  suggestions.begin(), suggestions.end(),
							  [&](const std::vector<char> elem)
							  {
								  return elem.size() != inputVector.size(); // put your condition here
							  }),
						  suggestions.end());
		// removes elements whose size doesnot match with input text

		for (int cc = 1; cc < inputVector.size(); cc++)
		{
			previousSuggestion = suggestions;
			// for saving suggestions one step before

			suggestions.erase(std::remove_if(suggestions.begin(), suggestions.end(),
											 [&](const std::vector<char> elem)
											 {
												 return !compare(elem.at(cc), inputVector.at(cc));
											 }),
							  suggestions.end());
			if (suggestions.size() == 0)
			{
				suggestions = previousSuggestion;
				cc++;
				if (cc < inputVector.size())
				{
					suggestions.erase(std::remove_if(suggestions.begin(), suggestions.end(),
													 [&](const std::vector<char> elem)
													 {
														 //  std::cout << "elem =" << elem.at(cc) << "    input =" << inputVector.at(cc) << std::endl;
														 return !compare(elem.at(cc), inputVector.at(cc));
													 }),
									  suggestions.end());
					if (suggestions.size() == 0)
					{
						suggestions = previousSuggestion;
						break;
					}
					previousSuggestion = suggestions;
				}
				else
				{
					break;
				}
			}
		}
		if (suggestions.size() > 5)
		{
			previousSuggestion = suggestions;
			suggestions.erase(std::remove_if(suggestions.begin(), suggestions.end(), [&](const std::vector<char> elem)
											 { return !compare(elem.at(0), inputVector.at(0)); }),
							  suggestions.end());

			if (suggestions.size() == 0)
				suggestions = previousSuggestion;
			if (suggestions.size() > 5)
				suggestions.erase(suggestions.begin() + 5, suggestions.end());
		}
		if (suggestions.size() == 0)
		{
			suggestions = previousSuggestion;
		}
		displayVectorDeque(suggestions);
	}
	else
	{
		std::cout << "Found your word at the dictionary\n";
		std::cout << "Here are the list of words near this one " << std::endl;
		displayCharVector(*(foundIterator - 1));
		displayCharVector(*(foundIterator + 1));
	}
	return 0;
}
// speed changed from (5-6) to (6-8)

File: Js/auto.js
let input = "nuxt";
let words = require("./words");
let compare = (a, b) => {
    if (a.toUpperCase() === b.toUpperCase()) {
        return true;
    }
    return false;
}
let wordFound = words.indexOf(input);
if (wordFound === -1) {
    console.log("No matching words found ");
    console.log("Suggestions for you are \n");
    let suggestions, previousSuggestion;
    suggestions = words;
    suggestions = suggestions.filter(currentElem => {
        return currentElem.length === input.length;
    });
    // removes elements whose size doesnot match with input text
    for (let curr = 0; curr < input.length; curr++) {
        previousSuggestion = suggestions;
        // for saving suggestions one step before

        suggestions = suggestions.filter(currentElem => {
            return compare(currentElem.charAt(curr), input.charAt(curr));
        });
        console.log("Suggestions after filtering first level ", curr, input[curr], suggestions)
        // Here the filtering is opposite than that of c++
        // filtering words whose character matches

        if (suggestions.length === 0) {
            suggestions = previousSuggestion;
            curr++;
            console.log("Suggestions after filtering first level ", curr, input[curr], suggestions)

            if (curr < input.length) {
                suggestions = suggestions.filter(currentElem => {
                    return compare(currentElem.charAt(curr), input.charAt(curr));
                });
                console.log("Where the length < 0 ", suggestions);
                if (suggestions.length === 0) {
                    suggestions = previousSuggestion;
                    break;
                }
                previousSuggestion = suggestions;
            } else {
                break;
            }
        }
    }
    if (suggestions.length > 5) {
        previousSuggestion = suggestions;
        suggestions = suggestions.filter(currentElem => {
            return compare(input.charAt(0), currentElem.charAt(0));
        });
        if (suggestions.length === 0) {
            suggestions = previousSuggestion;
        }
        if (suggestions.length > 5) {
            suggestions = suggestions.slice(0, 5);
            // for slicing the first 5 members
        }
    }
    if (suggestions.length === 0) {
        suggestions = previousSuggestion;
    }

    console.log(suggestions);

} else {
    console.log("Your word was found \n Here are words near it");
    console.log(words[wordFound - 1], words[wordFound + 1]);
}
// It gives more importance to words at first